### PR TITLE
Address collector

### LINKED ISSUE #1254: Create AutoCompleteController
AccountsController::autoComplete should be moved into its own controller as it is not related to a specific account

### PR DESCRIPTION
**Pull request by ChristophWurst**
Adds all addresses of sent mails (to, cc, bcc) to a db table, so that those addresses can then be used for auto-completion.

This is a very simple implementation, it only saves the bare email address. Adding more information (label, recipient name) can be done in another PR as that might need some advanced update-logic.

We should also think about a way for users to delete one of those entries…

TODO:
- [x] fix sql matching
- [x] update/add unit tests

implements #879
fixes #1254

### CODE REVIEW

File: appinfo/database.xml [hunk 1]
```diff
@@ -116,4 +116,44 @@
...
+				<name>mail_collected_addr_userid_index</name>
+				<field>
+					<name>user_id</name>
+					<sorting>ascending</sorting>
+				</field>
+			</index>
+			<index>
+				<name>mail_collected_addr_email_index</name>
+				<field>
+					<name>user_id</name>
```
Comment (ChristophWurst) [inline appinfo/database.xml]: this should be 'email'

File: lib/db/collectedaddressmapper.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+	/**
+	 * Find email addresses by query string
+	 *
+	 * @param string $userId
+	 * @param string $query
+	 * @return CollectedAddress[]
+	 */
+	public function findMatching($userId, $query) {
+		$sql = 'SELECT * FROM *PREFIX*mail_collected_addresses WHERE `user_id` = ? AND `email` ILIKE ?';
```
Comment (ChristophWurst) [inline lib/db/collectedaddressmapper.php]: @DeepDiver1975 I would like query as ``AND `email` ILIKE *?*`` but that's throwing an exception. Any idea if that's somehow possible (with owncloud)?

File: lib/db/collectedaddressmapper.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		$params = [
+			$userId,
+			$query
+		];
+		return $this->findEntities($sql, $params);
+	}
+
+	public function exists($userId, $email) {
+		$sql = 'SELECT * FROM *PREFIX*mail_collected_addresses WHERE `user_id` = ? AND `email` = ?';
```
Comment (ChristophWurst) [inline lib/db/collectedaddressmapper.php]: - [ ] also use ILIKE here

File: lib/db/collectedaddressmapper.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/db/collectedaddressmapper.php]: wildcard character is % - so you can pass in as parameter :

```php
'%' . $query . '%'

Comment (ChristophWurst) [inline lib/db/collectedaddressmapper.php]: ok thanks. I added the (wrong) wildcards to the prepared sql instead of the actual parameters :see_no_evil:

**Comment by ChristophWurst** [comment]
Dang it. Some test cases still fail

**Comment by ChristophWurst** [comment]
This is weird:
```
exception 'OCP\AppFramework\QueryException' with message 'Could not resolve OCP\IDBConnection! Class can not be instantiated' in /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php:80

Stack trace:

#0 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(98): OC\AppFramework\Utility\SimpleContainer->resolve('OCP\\IDBConnecti...')

#1 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(59): OC\AppFramework\Utility\SimpleContainer->query('OCP\\IDBConnecti...')

#2 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(78): OC\AppFramework\Utility\SimpleContainer->buildClass(Object(ReflectionClass))

#3 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(98): OC\AppFramework\Utility\SimpleContainer->resolve('OCA\\Mail\\Db\\Col...')

#4 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(59): OC\AppFramework\Utility\SimpleContainer->query('OCA\\Mail\\Db\\Col...')

#5 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(78): OC\AppFramework\Utility\SimpleContainer->buildClass(Object(ReflectionClass))

#6 /home/travis/build/owncloud/core/lib/private/appframework/utility/simplecontainer.php(98): OC\AppFramework\Utility\SimpleContainer->resolve('OCA\\Mail\\Servic...')

#7 /home/travis/build/owncloud/core/apps/mail/lib/account.php(90): OC\AppFramework\Utility\SimpleContainer->query('OCA\\Mail\\Servic...')

#8 /home/travis/build/owncloud/core/apps/mail/tests/imap/abstracttest.php(71): OCA\Mail\Account->__construct(Object(OCA\Mail\Db\MailAccount))

#9 [internal function]: OCA\Mail\Tests\Imap\AbstractTest::setUpBeforeClass()

#10 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestSuite.php(673): call_user_func(Array)

#11 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestSuite.php(716): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))

#12 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/TestRunner.php(398): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))

#13 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/Command.php(152): PHPUnit_TextUI_TestRunner->doRun(Object(PHPUnit_Framework_TestSuite), Array)

#14 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/Command.php(104): PHPUnit_TextUI_Command->run(Array, true)

#15 /home/travis/.phpenv/versions/5.4.37/bin/phpunit(722): PHPUnit_TextUI_Command::main()

#16 {main}
```
It's only reproducible on php5.4 and I have no idea what might cause that exception. @DeepDiver1975 is there any difference between php5.4 and php5.6 in terms of owncloud dependency injection?

**Comment by Gomez** [comment]
I dropping 5.4 a option for us (i would say yes)? The news app dropped it and 5.4 is EOL.

File: lib/account.php [hunk 4]
```diff
@@ -83,6 +87,7 @@ public function __construct(MailAccount $account) {
 		$this->crypto = \OC::$server->getCrypto();
 		$this->config = \OC::$server->getConfig();
 		$this->memcacheFactory = \OC::$server->getMemcacheFactory();
+		$this->addressCollector = \OC::$server->query('OCA\Mail\Service\AutoCompletion\AddressCollector');
```
Comment (DeepDiver1975) [inline lib/account.php]: @ChristophWurst this is the reason why this is falling apart.

In previous versions of owncloud you cannot query the server container for services which are located in an app ... let me fix that ...

**Comment by ChristophWurst** [comment]
@DeepDiver1975 thanks for fixing

@owncloud/mail @ErikPel @jospoortvliet review this please ;-)

**Comment by jancborchardt** [comment]
Nice @ChristophWurst, awesome on that functionality! :)

Do Gmail and others also collect addresses of people you receive emails _from_? (As well as people in cc there.)

And of course as you mention – saving the name would be cool, but for a follow-up PR.

**Comment by enoch85** [comment]
I know this is for another PR, but I'd love this to be synced with my CardDav as well so that when I write on mobile, the address is there as well.

**Comment by ChristophWurst** [comment]
> I dropping 5.4 a option for us (i would say yes)? The news app dropped it and 5.4 is EOL.

If we want to become an "official" app we have to support all platforms supported by core, see https://doc.owncloud.org/server/8.2/developer_manual/app/publishing.html#official

**Comment by Gomez** [comment]
@ChristophWurst Cool! Checked & Tested

@enoch85 Would you open a issue? This PR will be closed.

**Pull request merged**